### PR TITLE
Extend nGraph for operation CTCLoss

### PR DESCRIPTION
**Pull request by rkazants**
Signed-off-by: Roman Kazantsev <roman.kazantsev@intel.com>

### CODE REVIEW

**Comment by rkazants** [comment]
@nshchego, please proceed with development of CTCLoss ref implementation within this branch. I gave you a contributor's permission and you need to accept it.

File: ngraph/test/type_prop/ctc_loss.cpp [hunk 1]
```diff
@@ -0,0 +1,320 @@
...
+    auto logit_length = make_shared<op::Parameter>(element::i32, Shape{10});
+    auto labels = make_shared<op::Parameter>(element::i32, Shape{10});
+    auto label_length = make_shared<op::Parameter>(element::i32, Shape{10});
+    auto blank_index = make_shared<op::Parameter>(element::i32, Shape{});
+
+    try
+    {
+        // create CTCLoss node
+        auto ctc_loss =
+            make_shared<op::v4::CTCLoss>(logits, logit_length, labels, label_length, blank_index);
```
Comment (ilyachur) [inline ngraph/test/type_prop/ctc_loss.cpp:189]: Please use EXPECT_THROW instead of try ... catch

File: ngraph/src/ngraph/op/ctc_loss.cpp [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+                labels_pshape[1].compatible(time_steps),
+                "The second dimension of labels must be equal to the second dimension ",
+                "of logits. Got: ",
+                labels_pshape[1],
+                " and: ",
+                time_steps);
+        }
+        else if (labels_pshape[1].is_static())
+        {
+            time_steps = labels_pshape.to_shape()[1];
```
Comment (ilyachur) [inline ngraph/src/ngraph/op/ctc_loss.cpp:178]: Looks like we don't need this check

File: ngraph/src/ngraph/op/ctc_loss.cpp [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+                label_length_pshape[0].compatible(batch_size),
+                "The first dimension of label length must be equal to the first dimension ",
+                "of the logits, the logit length and labels. Got: ",
+                label_length_pshape[0],
+                " and: ",
+                batch_size);
+        }
+        else if (label_length_pshape[0].is_static())
+        {
+            batch_size = label_length_pshape.to_shape()[0];
```
Comment (ilyachur) [inline ngraph/src/ngraph/op/ctc_loss.cpp:198]: Maybe at the end of this code block we need add a check
```
NODE_VALIDATION_CHECK(
                this,
                label_length_pshape[0].compatible(batch_size),
                "The first dimension of label length must be equal to the first dimension ",
                "of the logits, the logit length and labels. Got: ",
                label_length_pshape[0],
                " and: ",
                batch_size);
```

Comment (nshchego) [reply ngraph/src/ngraph/op/ctc_loss.cpp:181]: Fixed

File: ngraph/src/ngraph/op/ctc_loss.cpp (hunk 2 above)
Comment (nshchego) [reply ngraph/src/ngraph/op/ctc_loss.cpp]: Fixed

File: ngraph/test/type_prop/ctc_loss.cpp (hunk 1 above)
Comment (nshchego) [reply ngraph/test/type_prop/ctc_loss.cpp]: fixed

Comment (ilyachur) [reply ngraph/test/type_prop/ctc_loss.cpp:189]: Oh, ok. Do we really need to check the error message?

Comment (nshchego) [reply ngraph/test/type_prop/ctc_loss.cpp:189]: I suppose it was made to check certain scenario: such inputs lead to appropriate message.

File: ngraph/src/ngraph/op/ctc_loss.cpp [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+                          "Expected a scalar for blank index. Got: ",
+                          blank_index_pshape);
+
+    // check shapes of input tensors
+    size_t batch_size = 1;
+    bool is_batch_size_set = false;
+    size_t time_steps = 1;
+    bool is_time_steps_set = false;
+
+    if (logits_pshape.is_static())
```
Comment (GlebKazantaev) [inline ngraph/src/ngraph/op/ctc_loss.cpp:112]: Check that rank is static

File: ngraph/src/ngraph/op/ctc_loss.cpp [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+    {
+        if (logits_pshape[0].is_static())
+        {
+            batch_size = logits_pshape.to_shape()[0];
```
Comment (GlebKazantaev) [inline ngraph/src/ngraph/op/ctc_loss.cpp:116]:
```suggestion
            batch_size = logits_pshape[0].get_length();
```

Comment (GlebKazantaev) [reply ngraph/src/ngraph/op/ctc_loss.cpp:116]: The same comment for all other places.

File: ngraph/src/ngraph/op/ctc_loss.hpp [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+                    clone_with_new_inputs(const OutputVector& new_args) const override;
+
+                bool get_preprocess_collapse_repeated() const
+                {
+                    return preprocess_collapse_repeated_;
+                }
+                bool get_ctc_merge_repeated() const { return ctc_merge_repeated_; }
+                bool get_unique() const { return unique_; }
+            private:
+                bool preprocess_collapse_repeated_;
```
Comment (GlebKazantaev) [inline ngraph/src/ngraph/op/ctc_loss.hpp:70]: ngraph usually uses `m_<variable_name>` for class members

File: ngraph/src/ngraph/op/ctc_loss.cpp (hunk 4 above)
Comment (nshchego) [reply ngraph/src/ngraph/op/ctc_loss.cpp]: Fixed

File: ngraph/src/ngraph/op/ctc_loss.cpp (hunk 5 above)
Comment (nshchego) [reply ngraph/src/ngraph/op/ctc_loss.cpp]: Fixed

File: ngraph/src/ngraph/op/ctc_loss.hpp (hunk 6 above)
Comment (nshchego) [reply ngraph/src/ngraph/op/ctc_loss.hpp:70]: dldt follows to google style and ngraph as its part also should.

**Comment by nshchego** [comment]
> New op should be added to nGraph Python API.

I've created separate task for that

File: ngraph/src/ngraph/op/ctc_loss.hpp (hunk 6 above)
Comment (GlebKazantaev) [reply ngraph/src/ngraph/op/ctc_loss.hpp:70]: I do not agree but I would keep it for separate discussion.

Comment (ilyachur) [reply ngraph/src/ngraph/op/ctc_loss.hpp:70]: At the current moment nGraph uses LLVM code style.

**Pull request merged**